File: Recursion/leetcode_394/solution.py
class Solution:
    def decodeString(self, s: str) -> str:
        self.index = 0

        def decodeRecursive():
            current_string = []
            current_number = 0 

            while self.index < len(s):
                char = s[self.index]

                if '0' <= char <= '9':
                    current_number = current_number * 10 + int(char)

                elif char == '[':
                    self.index += 1
                    
                    nested_decoded_string = decodeRecursive() 

                    current_string.append(nested_decoded_string * current_number)
                    current_number = 0

                elif char == ']':
                    return "".join(current_string)

                else:
                    current_string.append(char)
                
                self.index += 1 
            
            return "".join(current_string)

        return decodeRecursive()

File: Recursion/leetcode_394/solution.js
/**
 * @param {string} s
 * @return {string}
 */
var decodeString = function(s) {
    let index = 0;

    function decodeRecursive() {
        let currentString = [];
        let currentNumber = 0;

        while (index < s.length) {
            const char = s[index];

            if (char >= '0' && char <= '9') {
                currentNumber = currentNumber * 10 + Number(char);
            } else if (char === '[') {
                index++;
                
                const nestedDecodedString = decodeRecursive(); 

                for (let i = 0; i < currentNumber; i++) {
                    currentString.push(nestedDecodedString);
                }
                currentNumber = 0;
            } else if (char === ']') {
                return currentString.join('');
            } else {
                currentString.push(char);
            }

            index++; 
        }

        return currentString.join('');
    }

    return decodeRecursive();
};